builtin/plugins/dposv3/dpos.go

```diff
@@ -736,6 +737,54 @@ func (c *DPOS) Unbond(ctx contract.Context, req *UnbondRequest) error {
 	return c.emitDelegatorUnbondsEvent(ctx, delegation)
 }
 
+func (c *DPOS) UnbondAll(ctx contract.Context, req *UnbondAllRequest) error {
+	if !ctx.FeatureEnabled(features.DPOSVersion3_7, false) {
+		return errors.New("DPOS v3.7 is not enabled")
+	}
+
+	sender := ctx.Message().Sender
+	ctx.Logger().Info("DPOSv3", "UnbondAll", "sender", sender, "request", req)
+
+	state, err := LoadState(ctx)
+	if err != nil {
+		return err
+	}
+
+	// ensure that function is only executed when called by oracle
+	if state.Params.OracleAddress == nil || sender.Compare(loom.UnmarshalAddressPB(state.Params.OracleAddress)) != 0 {
+		return logDposError(ctx, errOnlyOracle, req.String())
+	}
+
+	delegationIndexes, err := loadDelegationList(ctx)
+	if err != nil {
+		return logDposError(ctx, errors.New("Existing delegation not in BONDED state."), req.String())
+	}
+
+	for _, di := range delegationIndexes {
+		delegation, err := GetDelegation(ctx, di.Index, *di.Validator, *di.Delegator)
+		if err == contract.ErrNotFound {
+			return logDposError(ctx, errors.New(fmt.Sprintf("delegation not found: %s %s", di.Validator, di.Delegator)), req.String())
+		} else if err != nil {
+			return errors.Wrap(err, "failed to load delegation")
+		}
+
+		if delegation.State != BONDED {
+			return logDposError(ctx, errors.New("Existing delegation not in BONDED state."), req.String())
```

Comment:
Shouldn't return an error in this case, check if `UpdateAmount` different to `Amount` - if it is set `UpdateAmount` to match `Amount`.